### PR TITLE
[RTM] Store the request token as cookie instead of the session

### PR DESCRIPTION
**Pull request by ausi**
Use a [double submit cookie](https://www.owasp.org/index.php/Cross-Site_Request_Forgery_(CSRF)_Prevention_Cheat_Sheet#Double_Submit_Cookie) instead of storing the request token in the session of the user.

### CODE REVIEW

File: src/Csrf/MemoryTokenStorage.php [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+    {
+        $this->tokens = $tokens;
+    }
+
+    /**
+     * Get all active tokens.
+     *
+     * @return array
+     */
+    public function getSaveTokens()
```
Comment (aschempp) [inline src/Csrf/MemoryTokenStorage.php]: I don't like this method name, shouldn't it be `getActiveTokens` if the internal variable is called `$activeTokens`? Also, I would prefer to name that `usedTokens` 😁 

The method can be simplified like this:

```php
return array_intersect_key($this->tokens, $this->activeTokens);
```

File: src/EventListener/CsrfTokenCookieListener.php [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    private $cookiePrefix;
+
+    /**
+     * Constructor.
+     *
+     * @param MemoryTokenStorage $tokenStorage
+     * @param int                $cookieLifetime
+     * @param string             $cookiePrefix
+     */
+    public function __construct(MemoryTokenStorage $tokenStorage, int $cookieLifetime = 86400, string $cookiePrefix = 'csrf_')
```
Comment (aschempp) [inline src/EventListener/CsrfTokenCookieListener.php]: Does this have to be a `MemoryTokenStorage`? Shouldn't we expect the interface?

File: src/EventListener/CsrfTokenCookieListener.php [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+     * @param ParameterBag $cookies
+     *
+     * @return array
+     */
+    private function getTokensFromCookies(ParameterBag $cookies)
+    {
+        $tokens = [];
+
+        foreach ($cookies as $key => $value) {
+            if (strncmp($key, $this->cookiePrefix, strlen($this->cookiePrefix)) === 0) {
```
Comment (aschempp) [inline src/EventListener/CsrfTokenCookieListener.php]: `strpos($this->cookiePrefix, $key) === 0` ?

File: src/Resources/config/listener.yml [hunk 4]
```diff
@@ -161,3 +161,11 @@ services:
...
             - { name: kernel.event_listener, event: kernel.request, method: onKernelRequest }
             - { name: kernel.event_listener, event: kernel.response, method: onKernelResponse }
+
+    contao.listener.csrf_token_cookie:
+        class: Contao\CoreBundle\EventListener\CsrfTokenCookieListener
+        arguments:
+            - "@contao.csrf.token_storage"
+        tags:
+            - { name: kernel.event_listener, event: kernel.request, method: onKernelRequest, priority: 20 }
+            - { name: kernel.event_listener, event: kernel.response, method: onKernelResponse, priority: 0 }
```
Comment (aschempp) [inline src/Resources/config/listener.yml]: you don't need to set a priority if it's `0`.

File: src/EventListener/CsrfTokenCookieListener.php (hunk 2 above)
Comment (ausi) [reply src/EventListener/CsrfTokenCookieListener.php]: We call the `initialize()` method on it, so we cannot use the interface.

File: src/EventListener/CsrfTokenCookieListener.php (hunk 3 above)
Comment (ausi) [reply src/EventListener/CsrfTokenCookieListener.php]: `strncmp()` should be way faster IMO.

File: src/Csrf/MemoryTokenStorage.php (hunk 1 above)
Comment (ausi) [reply src/Csrf/MemoryTokenStorage.php]: Done in c3f6afb94f567a587c763f1a32e84d8ecd29bdcb

File: src/Resources/config/listener.yml (hunk 4 above)
Comment (ausi) [reply src/Resources/config/listener.yml]: Done in d874349ecc242ae71e3fc096da15369a9c281aa3

File: src/EventListener/CsrfTokenCookieListener.php [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    {
+        if (!$event->isMasterRequest()) {
+            return;
+        }
+
+        $cookieLifetime = $this->cookieLifetime ? $this->cookieLifetime + time() : 0;
+        $isSecure = $event->getRequest()->isSecure();
+
+        foreach ($this->tokenStorage->getUsedTokens() as $key => $value) {
+            $event->getResponse()->headers->setCookie(
```
Comment (aschempp) [inline src/EventListener/CsrfTokenCookieListener.php]: I think you should use `deleteCookie` if the value is `null`. This way Symfony handles the unset case.

Comment (ausi) [reply src/EventListener/CsrfTokenCookieListener.php]: deleteCookie needs the same arguments as setCookie, except value and lifetime. Using deleteCookie would make a block of duplicate code.

Comment (aschempp) [reply src/EventListener/CsrfTokenCookieListener.php]: but it would move the delete logic to the header bag instead of our class, right?

Comment (ausi) [reply src/EventListener/CsrfTokenCookieListener.php]: If you refer to `$value === null ? 1 : $cookieLifetime` as the delete logic, then yes.

**Pull request merged**